src/bucket/BucketSnapshotManager.cpp

```diff
@@ -61,9 +64,13 @@ BucketSnapshotManager::recordBulkLoadMetrics(std::string const& label,
 
 void
 BucketSnapshotManager::maybeUpdateSnapshot(
-    std::unique_ptr<BucketListSnapshot const>& snapshot) const
+    std::unique_ptr<BucketListSnapshot const>& snapshot,
+    std::map<uint32_t, std::unique_ptr<BucketListSnapshot const>>&
```

Comment:
What is the purpose of using unique ptrs here, given that you're anyway making copies?

Reply:
This function is called by a background thread to see if their local BucketList snapshot is up to date.

The interface is such that each thread should maintain a single BucketList snapshot. The thread is the sole owner of that snapshot, thus `unique_ptr`. In order to make sure that each snapshot has an independent set of io streams, we always construct new snapshots by copy. In order to indicate that each thread should maintain a single uniquely owned snapshot, the `BucketSnapshotManager` returns a `unique_ptr` whenever a thread requests a new snapshot.

Finally, it is the responsibility of the `BucketSnapshotManager` to update other threads snapshots via this function. While we could do this with a copy assignment overload, I personally think swapping a unique_ptr with a freshly copy constructed object provides the safest and cleanest approach. I'm worried about bugs where you could have a snapshot that's partially at ledger n and partially at ledger n + 1, so I prefer always just constructing a new object and swapping the pointer.

Reply:
The code would've been simpler if we could just assign the map (not sure why does that require an overload for copy assignment TBH), but it's not a big deal, so if that's a conscious decision we can leave the ptr as is.